_posts/2017-11-23-totw-126.md

```diff
@@ -0,0 +1,117 @@
+
+1.  When possible, ownership is best expressed in the type system. This allows
+    reviewers to verify correctness (absence of leaks and of double-deletes)
+    almost entirely by local inspection. (In code that is exceptionally
+    performance sensitive, this may be excused: while cheap, passing
+    `std::unique_ptr` across function boundaries by value has non-zero overhead
+    because of ABI constraints. That's rarely important enough to justify
+    avoiding it.)
+
+2.  Somewhat like the reasoning for preferring `push_back` over `emplace` ([TotW
+    112](/tips/112)), `absl::make_unique` directly expresses the intent and can
+    only do one thing (do the allocation with a public constructor, returning a
+    `std::unique_ptr` of the specified type). There's no type conversion or
+    hidden behavior. `absl::make_unique` does what it says on the tin.
+
+3.  The same could be achieved with `std::unique_ptr<T> my_t(new T(args));` but
+    that is redundant (repeating the type name `T`) and for some people there's
+    value in minimizing calls to `new`. More on this in \#5.
+
+4.  If all allocations are handled via `absl::make_unique` or factory calls,
+    that leaves `absl::WrapUnique` for the implementation of those factory
+    calls, for code interacting with legacy methods that don't rely on
+    `std::unique_ptr` for ownership transfer, and for rare cases that need to
+    dynamically allocate with aggregate initialization (`absl::WrapUnique(new
+    MyStruct{3.141, "pi"})`). In code review it's easy to spot the
+    `absl::WrapUnique` calls and evaluate "does that expression look like an
+    ownership transfer?" Usually it's obvious (for example, it's some factory
+    function). When it's not obvious, we need to check the function to be sure
+    that it's actually a raw-pointer ownership transfer.
+
+5.  If we instead rely mostly on the constructors of `std::unique_ptr`, we see
+    calls like: \
+    `std::unique_ptr<T> foo(Blah());` \
+    `std::unique_ptr<T> bar(new T());` \
+    It takes only a moment's inspection to see that the latter is safe (no leak,
+    no double-delete). The former? It depends: if `Blah()` is returning a
+    `std::unique_ptr`, it's fine, though in that case it would be more obviously
+    safe if written as \
+    `std::unique_ptr<T> foo = Blah();` \
+    If `Blah()` is returning an ownership-transferred raw pointer, that's also
+    fine. If `Blah()` is returning just some random pointer (no transfer), then
+    there's a problem. Reliance on `absl::make_unique` and `absl::WrapUnique`
+    (avoiding the constructors) provides an additional visual clue for the
+    places where we have to worry (calls to `absl::WrapUnique`, and only those).
+
+How should we choose which to use?
+
+1.  By default, use `absl::make_unique` (or `std::make_shared` for the rare
+    cases where shared ownership is appropriate) for dynamic allocation. For
+    example, instead of \
+    `std::unique_ptr<T> bar(new T());` \
```

Comment:
The \ here don't do what needs to be done. Unfortunately, it needs to be some combination of \`\`\` (flush with the lefthand side) and indenting (for the inline text). See below (whitespace is significant here, though note that I did not escape the text below (view it raw to see what I mean).

1.  By default, use `absl::make_unique` (or `std::make_shared` for the rare
    cases where shared ownership is appropriate) for dynamic allocation. For
    example, instead of:
```
std::unique_ptr<T> bar(new T());
```
    write
```
auto bar = absl::make_unique<T>();
```
    and instead of
```
bar.reset(new T());
```
    write
```
bar = absl::make_unique<T>();
```

Reply:
I don't mind it inline - is that OK?